cloudstate/cloudstate.go

```diff
@@ -0,0 +1,249 @@
+//
+// Copyright 2019 Lightbend Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloudstate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/cloudstateio/go-support/cloudstate/protocol"
+	"github.com/golang/protobuf/descriptor"
+	"github.com/golang/protobuf/proto"
+	filedescr "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+	"log"
+	"net"
+	"os"
+	"runtime"
+)
+
+const (
+	SupportLibraryVersion = "0.1.0"
+	SupportLibraryName    = "cloudstate-go-support"
+)
+
+// CloudState is an instance of a CloudState User Function
+type CloudState struct {
+	server                   *grpc.Server
+	entityDiscoveryResponder *EntityDiscoveryResponder
+	eventSourcedHandler      *EventSourcedHandler
+}
+
+// NewCloudState returns a new CloudState instance.
+func NewCloudState(options *Options) *CloudState {
+	cs := &CloudState{
+		server:                   grpc.NewServer(),
+		entityDiscoveryResponder: NewEntityDiscoveryResponder(options),
+		eventSourcedHandler:      NewEventSourcedHandler(),
+	}
+	protocol.RegisterEntityDiscoveryServer(cs.server, cs.entityDiscoveryResponder)
+	log.Println("RegisterEntityDiscoveryServer")
+	protocol.RegisterEventSourcedServer(cs.server, cs.eventSourcedHandler)
+	log.Println("RegisterEventSourcedServer")
+	return cs
+}
+
+// Options go get a CloudState instance configured.
+type Options struct {
```

Comment:
> It's more typical to call this struct Config.

Agreed.
>As these values are not options that I can tell - they are required, right?

Not according to the protocol defined in [cloudstate/entity.proto](https://github.com/cloudstateio/cloudstate/blob/master/protocols/protocol/cloudstate/entity.proto#L132). I followed the requirements there. But even if some of them are optional and some mandatory, naming it Config makes sense, right?

- [ ] call this struct `Config` instead of `Options`
- [ ] consolidate other options possibly required by the API here and apply [functional option pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis).

Reply:
Ah interesting. I thought the service_name was required, at least. Since this is used by the proxy to forward commands/events to other services, right?